src/config/hypre_macros_misc.m4

```diff
@@ -220,31 +220,31 @@ if test "x${hypre_user_chose_cflags}" = "xno"
 then
    case `basename ${CC}` in
       gcc|mpigcc|mpicc)
-        CFLAGS="-g -Wall"
+        CFLAGS="-O0 -g -Wall"
```

Comment:
I don't think we should add the '-O0' to '--enable-debug'.  I think this should just turn on debugging, not modify the optimization setting.  Why do you think this is a good thing to always do here?  Sometimes it is good to debug with different optimization settings (the bug may be caused by the optimization setting).

Reply:
Good catch. I didn't see this change since I thought they are all the "format changes". We also have `-O0` in CUFLAG with `--enable-debug`, which may not be a good thing either, see https://github.com/hypre-space/hypre/blob/70e9dec6bda36e6a60b98b58eff0656bb840d0d7/src/config/configure.in#L2543. The reason I added was optimization sometimes can remove information used by debuggers. I usually use debuggers like gdb with `--enable-debug`.

Reply:
Because the compiler is free to choose the optimization level if not specified by the user. In many cases, the optimization level chosen by default is `-O2`. However, when debugging with gdb, `-O2` generally hides the value of local variables, which doesn't help debugging.

The point of a bug being related to a particular optimization flag is good. If that is the case, the user could do `--with-extra-CFLAGS="-O2"`, for example, and `O2` would have the priority, thus bringing back the bug again

Reply:
Let's get a consensus on this question via teams.  For me, I use '--enable-debug' all the time and almost never have to additionally change the optimization flag (when I have done that in the past, it was typically because the optimization setting was causing the bug in the first place).  I usually use totalview, not gdb.  So, this will change the behavior of '--enable-debug' for me and I'm not sure if it will be for better or worse until I start using it.  My guess is I won't notice anything at all.

Reply:
Whatever we decide, we should be consistent.

Reply:
> the compiler is free to choose the optimization level [...] In many cases [...] default is -O2.

I did some research to clarify this and found the following defaults:
| Compiler | Opt. level |
|----------|---------------|
| `icx`/`icc` | `-O2` |
| `xlc` | `-O2` |
| `gcc` | `-O0` |
| `clang` | `-O0` |
| `msvc` | `-O0` |

I think it would be good to have an expected behavior with `--enable-debug`, either by assuring `-O0` or `-O2`.

Also, `CMake` sets `-O0` with its `-DCMAKE_BUILD_TYPE=Debug` option. So, I think it makes sense to get the same with configure's `--enable-debug` for consistency.

Reply:
This seems reasonable to me.  @liruipeng , what do you think?